state-sync/state-sync-v2/state-sync-driver/src/bootstrapper.rs

```diff
@@ -0,0 +1,848 @@
+                        data_notification.notification_id,
+                        epoch_ending_ledger_infos,
+                    )
+                    .await?;
+                }
+                DataPayload::TransactionsWithProof(transactions_with_proof) => {
+                    let payload_start_version = transactions_with_proof.first_transaction_version;
+                    self.process_transaction_or_output_payload(
+                        data_notification.notification_id,
+                        Some(transactions_with_proof),
+                        None,
+                        payload_start_version,
+                    )
+                    .await?;
+                }
+                DataPayload::TransactionOutputsWithProof(transaction_outputs_with_proof) => {
+                    let payload_start_version =
+                        transaction_outputs_with_proof.first_transaction_output_version;
+                    self.process_transaction_or_output_payload(
+                        data_notification.notification_id,
+                        None,
+                        Some(transaction_outputs_with_proof),
+                        payload_start_version,
+                    )
+                    .await?;
+                }
+                _ => {
+                    return self
+                        .handle_end_of_stream_or_invalid_payload(data_notification)
+                        .await
+                }
+            }
+        }
+    }
+
+    /// Fetches all epoch ending ledger infos (from the current epoch to the
+    /// maximum that can be found by the data streaming service).
+    async fn fetch_epoch_ending_ledger_infos(
+        &mut self,
+        global_data_summary: &GlobalDataSummary,
+    ) -> Result<(), Error> {
+        // Verify the waypoint can be satisfied
+        self.verify_waypoint_is_satisfiable(global_data_summary)?;
+
+        // Get the highest advertised epoch that has ended
+        let highest_advertised_epoch_end = global_data_summary
+            .advertised_data
+            .highest_epoch_ending_ledger_info()
+            .ok_or_else(|| {
+                Error::AdvertisedDataError(
+                    "No highest advertised epoch end found in the network!".into(),
+                )
+            })?;
+
+        // Fetch the highest epoch end known locally
+        let highest_known_ledger_info = self.get_highest_known_ledger_info()?;
+        let highest_known_ledger_info = highest_known_ledger_info.ledger_info();
+        let highest_local_epoch_end = if highest_known_ledger_info.ends_epoch() {
+            highest_known_ledger_info.epoch()
+        } else if highest_known_ledger_info.epoch() > 0 {
+            highest_known_ledger_info
+                .epoch()
+                .checked_sub(1)
+                .ok_or_else(|| {
+                    Error::IntegerOverflow("The highest local epoch end has overflown!".into())
+                })?
```

Comment:
FYI, since you check the condition above, it will never fail.

Reply:
Yeah, I've given up reasoning about which checked arithmetic operations are genuine and those which aren't. I'm basically programming them like a robot now. Once this things all lands, we should figure out the right direction for all of these in our code. 😄